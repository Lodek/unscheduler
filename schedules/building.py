
class Building():
    """ Abstraction for a building plan from Archicad.
    A building plan can be thought of as a group of areas and stories. """
    
    def __init__(self):
        self.model = ''
        self._stories = []
        self.area_comp = 0.0
        self.area_ncomp = 0.0
        self._calc_net_area()

    @classmethod
    def get_from_text(cls, text):
        """ Construct a Building object from a text file
        It parses the text file, retrieves the Story objects associated to it
        and does the nescessary computations to initialize the Building object """
        pass

    @property
    def stories(self):
        """ Returns an iterator that iterates over Story objects in Building """
        pass

    def _calc_net_area(self):
        """ Calculate the net areas for Building object"""
        for story in self._stories:
            self.area_comp += story.area_comp
            self.area_ncomp += story.area_ncomp
    
    def __getitem__(self):
        """ Returns the Story object by index """
        pass

    def __len__(self):
        """ Number of stories in building """
        return len(self._stories)
    
    def __repr__(self):
        str = 'Model: {}, Stories: {}\nArea_C: {}m2\tArea_NC {}m2'
        str.format(self.model, len(self), self.area_comp, self.area_ncomp)
        return str
        
