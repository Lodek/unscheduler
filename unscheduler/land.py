
class Land():
    """ Base class for a piece of Land. 
    Land has attributes for its areas, coeficients and rates """
    def __init__(self):
        self.area_net = 0.0
        self.area_perm = 0.0
        self.area_comp = 0.0
        self.area_ncomp = 0.0
        self.area_proj = 0.0
        self.coef_aprov = 0.0
        self.taxa_perm = 0.0
        self.taxa_ocp = 0.0

    def calc_coef_aprov(self):
        """ Calculates the coef aprov for the instance """
        coef = self.area_comp/self.area_net
        return coef

    def calc_taxa_perm(self):
        """ Calculates the Land's permeability rate """
        rate = self.area_perm / self.area_net
        return rate

    def calc_taxa_ocp(self):
        """ Calculates the Land's ocp rate """
        rate = self.area_proj / self.area_net
        return rate

    def calc_all(self):
        """ Updates all of the instance's rates and coeffs. """
        self.coef_aprov = self.calc_coef_aprov()
        self.taxa_perm = self.calc_taxa_perm()
        self.taxa_ocp = self.calc_taxa_ocp()

        
class Lot(Land):
    """ Abstraction for an Archicad land plot.
    Land is built from subplots and data intrinsic to it """

    def __init__(self, subplots, project_info):
        super().__init__()
        self.area_lot = 0.0
        self.area_useless = 0.0
        self.units = 0
        self.rec_ncov = 0.0
        self.rec_cov = 0.0
        self.rec_net = 0.0
        self.subplots = subplots
        for key, value in project_info.items():
            setattr(self, key, value)

    def calc_all(self):
        self.area_net = sum([subplot.area_net for subplot in self.subplots])
        self.area_perm = sum([subplot.area_perm for subplot in self.subplots])
        self.area_comp = sum([subplot.area_comp for subplot in self.subplots])
        self.area_ncomp = sum([subplot.area_ncomp for subplot in self.subplots])
        self.area_proj = sum([subplot.area_proj for subplot in self.subplots])
        super().calc_all()

    def __len__(self):
        """ Returns the ammount of subplots """
        return len(self.subplots)

class Subplot(Land):

    def __init__(self, id, name, area_net, building=None):
        super().__init__()
        self.id = id
        self.area_net = area_net
        self.name = name
        self.building = building
        
    def calc_all(self):
        self.area_comp = self.building.area_comp
        self.area_ncomp = self.building.area_ncomp
        self.area_proj = self.building.area_proj
        super().calc_all()
        
    def __repr__(self):
        str = 'id: {}, name: {}, building model: {} \nNet Area: {}'
        str = str.format(self.id, self.name, self.building.model, self.area_net)
        return str
