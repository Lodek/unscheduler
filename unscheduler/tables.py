""" This file contains all of the classes responsible for generating the formatted .tex file.
Each type of Table is a subclass of BaseBuilder """

from pathlib import Path

templates_path = Path(__file__).parent / '..' / 'templates'

class LaTexFactory():
    """
    From list of lists formats a latex file
    """
    @classmethod
    def _tabularfy(cls, matrix, headers=True, title=False):
        """Generate simple latex formatting for a latex tabular environment. 
        Returns the contents of the Latex tabular environment as a list of strings.
        title flag indicates whether or not matrix the first row of the matrix
        contais a "title".
        headers flag indicates whether or not the second row (first if there's
        no title) is a row of headers.
        """
        hline = r'\hline'
        tabular = []
        if title:
            tabular.append(cls._format_row(matrix.pop(0)))
        if headers:
            tabular.append(hline)
            tabular.append(cls._format_row(matrix.pop(0)))
        tabular.append(hline)
        tabular.extend([cls._format_row(row) for row in matrix])
        tabular.append(hline)
        return tabular

    @staticmethod
    def _format_row(row):
        """ Takes a list of strings under the assumption that each
        element in that list is a cell in a tabular row. 
        Returns a string with matching latex code for the row """
        code = ' & '.join(row) + r' \\'
        return code

    @classmethod
    def get_latex(cls, matrix, template, args=[], headers=True, title=False):
        """From matrix (which contains the body of the table as a list of lists,
        return string which is the corresponding matrix in a LaTeX format
        inserted into template. Optionally passes *args to format call on template"""
        tabular = '\n'.join(cls._tabularfy(matrix, headers, title))
        return template.format(*args + [tabular])

    
fmt_area = lambda s : '${:.2f}$ m$^2$'.format(s)
fmt_perc = lambda s : '${:.2f}$\%'.format(s)
fmt_float = lambda s : '${:.2f}$'.format(s)

class StoryTableFactory:
    """
    Factory that returns the representation of a Story as a LaTeX table
    """
    @classmethod
    def get_latex(cls, story):
        template_path = templates_path / 'story-template.tex'
        template = template_path.read_text()
        table = cls._build_table(story)
        tex = LaTexFactory.get_latex(table, template, title=True)
        return tex

    @staticmethod
    def _build_table(story):
        title = ['', story.name.upper(), '']
        headers = ['AREA COMP.', 'AREA NAO COMP.', 'TOTAL']
        body = [[fmt_area(story.area_comp), fmt_area(story.area_ncomp), fmt_area(story.total)]]
        return [title, headers] + body


class BuildingTableFactory:
    """
    Factory class that returns the representation of a building as a table in Tex
    """
    @classmethod
    def get_latex(cls, building):
        template_path = templates_path / 'building-template.tex'
        template = template_path.read_text()
        table = cls._build_table(building)
        tex = LaTexFactory.get_latex(table, template)
        return tex
        
    @staticmethod
    def _build_table(building):
        """ Generates tabular list for Building Table """
        headers = ['PAVIMENTO', 'AREA COMP.', 'AREA NAO COMP.']
        formatter = lambda s : [s.name.upper(), fmt_area(s.area_comp), fmt_area(s.area_ncomp)]
        body = [formatter(story) for story in building.stories + [building.super_story]]
        return [headers] + body


class SubplotAreasFactory:
    """

    """
    @classmethod
    def get_latex(cls, lot):
        template_path = templates_path / 'subplot-areas-template.tex'
        template = template_path.read_text()
        table = cls._build_table(lot)
        table_fmt = 'l' * len(table[0])
        args = [table_fmt]
        tex = LaTexFactory.get_latex(table, template, args=args, title=True)
        return tex
 
    @staticmethod
    def _build_table(lot):
        title = ['']
        for story in lot.super_building.all_stories():
            title.extend([story.name.upper(), ''])
        header = ['SUBLOTE']
        for _ in lot.super_building.all_stories():
            header.extend(['COMP', 'N COMP'])
        body = []
        for s in lot.subplots + [lot]:
            row = []
            row.append(s.name.upper())
            for story in s.super_building.all_stories():
                row.append(fmt_area(story.area_comp))
                row.append(fmt_area(story.area_ncomp))
            body.append(row)
        return [title, header] + body


class SubplotStatsFactory:
    """ SubplotTable generates the .tex file for the Subplot statistics table. """
    @classmethod
    def get_latex(cls, lot):
        template_path = templates_path / 'subplot-stats-template.tex'
        template = template_path.read_text()
        table = cls._build_table(lot)
        tex = LaTexFactory.get_latex(table, template, headers=False)
        return tex

    @classmethod
    def _build_table(cls, lot):
        headers = ['NOME', 'AREA PROJECAO', 'AREA SUBLOTE', 'TAXA OCUPACAO',
                   'COEF.APROVEITAMENTO', 'AREA PERMEAVEL', 'TAXA PERMEABILIDADE']
        body = [cls._gen_row(sub) for sub in lot.subplots + [lot]]
        return body

    @staticmethod
    def _gen_row(s):
        row = [s.name.upper()]
        row.append(fmt_area(s.area_proj))
        row.append(fmt_area(s.area_net))
        row.append(fmt_perc(s.taxa_ocp))
        row.append(fmt_float(s.coef_aprov))
        row.append(fmt_area(s.area_perm))
        row.append(fmt_perc(s.taxa_perm))
        return row

class LotStatsFactory:
    """ LotTable generates the .tex file for the Lot statistics table. """
    @classmethod
    def get_latex(cls, lot):
        template_path = templates_path / 'lot-stats-template.tex'
        template = template_path.read_text()
        table = cls._build_table(lot)
        tex = LaTexFactory.get_latex(table, template, title=True, headers=False)
        return tex

    def _build_table(lot):
        title = ['ESTATISTICAS', '']
        body = [['AREA DO LOTE', fmt_area(lot.area_lot)]]
        body.append(['AREA DO LOTE ATINGIDA', fmt_area(lot.area_useless)])
        body.append(['AREA DO LOTE REMANESCENTE', fmt_area(lot.area_net)])
        body.append(['AREA TOTAL COMPUTAVEL', fmt_area(lot.area_comp)])
        body.append(['AREA TOTAL NAO COMPUTAVEL', fmt_area(lot.area_ncomp)])
        body.append(['AREA PROJECAO', fmt_area(lot.area_proj)])
        body.append(['TAXA DE OCUPACAO', fmt_perc(lot.taxa_ocp)])
        body.append(['COEFICIENTE DE APROVEITAMENTO', fmt_float(lot.coef_aprov)])
        body.append(['AREA PERMEAVEL', fmt_area(lot.area_perm)])
        body.append(['COEFICIENTE PERMEABILIDADE', fmt_perc(lot.taxa_perm)])
        body.append(['NUMERO DE UNIDADES', '{} UN'.format(lot.units)])
        body.append(['RECREACAO DESCOBERTA', fmt_area(lot.rec_ncov)])
        body.append(['RECREACAO COBERTA', fmt_area(lot.rec_cov)])
        body.append(['RECREACAO TOTAL', fmt_area(lot.rec_net)])
        return [title] + body

