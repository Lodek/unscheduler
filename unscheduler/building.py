""" 
The module contains memes.
"""
import logging
import tables

logger = logging.getLogger(__name__)

class Building():
    """
    A building is a named collection of stories.
    """
    formatter = tables.BuildingFormatter()
    def __init__(self, model, stories):
        self.model = model
        self.stories = stories
        self.super_story = None
        acs = [story.area_comp for story in self.stories]
        ncs = [story.area_ncomp for story in self.stories]
        self.super_story = Story(-1, 'TOTAL', sum(acs), sum(ncs))
        self.area_proj = max(acs)
        logger.info(repr(self))
        logger.debug('Building from stories={}'.format(stories))

    @classmethod
    def get_null_building(cls):
        return cls('null', [Story.null_story()])

    @classmethod
    def get_super_building(cls, model, buildings):
        """Super building return a building from a sequence of buildings as opposed
        to a sequence of stories. Super building does a per story sum of areas.
        Super building's idiosyncrasy is that area_proj instead of being the max
        becomes the sum of the projection areas for each building"""
        if not buildings:
            return cls.get_null_building()
        tallest_building = sorted(buildings, key=len)[-1]
        story_names = [story.name for story in tallest_building.stories]
        total_area_proj = sum(building.area_proj for building in buildings)
        stories = []
        for i, name in enumerate(story_names):
            ac = sum(building[i].area_comp for building in buildings)
            nc = sum(building[i].area_ncomp for building in buildings)
            stories.append(Story(i, name, ac, nc))
        new_building = cls(model, stories)
        new_building.area_proj = total_area_proj
        return new_building
            
    @property
    def area_comp(self):
        return self.super_story.area_comp

    @property
    def area_ncomp(self):
        return self.super_story.area_ncomp

    @property
    def total(self):
        return self.super_story.total

    def write_latex(self, target_dir):
        target = target_dir / '{}.tex'.format(self.model)
        latex = self.formatter.format(self)
        with target.open('w') as f:
            f.write(latex)

    def all_stories(self):
        return self.stories + [self.super_story]
    
    def __getitem__(self, n):
        """Returns the nth Story if not present, returns an empty story"""
        try:
            value = self.stories[n]
        except IndexError:
            value = Story(n, 'out-of-range')
        return value

    def __len__(self):
        """ Number of stories in building """
        return len(self.stories)

    def __repr__(self):
        s = '{}: model={}; stories={}; area_comp={:.2f}; area_ncomp={:.2f};'
        return s.format(self.__class__, self.model, len(self.stories), self.area_comp, self.area_ncomp)



class Story():
    """
    Abstraction of a Story.
    A story is composed of a name, an id, net computable area and net non-computable area
    Each building has multiple stories
    """
    formatter = tables.StoryFormatter()
    def __init__(self, id, name, area_comp=0.0, area_ncomp=0.0):
        self.id = id
        self.name = name
        self.area_comp = area_comp
        self.area_ncomp = area_ncomp
        logger.debug(repr(self))
        

    @classmethod
    def null_story(cls):
        return Story(0, 'null')
    
    def add_area(self, area, category):
        """Adds given area to the matching category,
        area: float
        category: str ('NC' or 'C') """
        logger.debug('Add area call: area={}, category={}, self={}'.format(area, category, repr(self)))
        if category == 'NC':
            self.area_ncomp += area
        elif category == 'C':
            self.area_comp += area
        else:
            raise ValueError('Invalid category')

    @property
    def total(self):
        return self.area_comp + self.area_ncomp

    def write_latex(self, target_dir):
        target_path = target_dir / 'story-{}.tex'.format(self.id)
        latex = self.formatter.format(self)
    
    def __repr__(self):
        s = '{}: id={}; name={}; area_comp={:.2f}; area_ncomp={:.2f};'
        return s.format(self.__class__, self.id, self.name, self.area_comp, self.area_ncomp)
