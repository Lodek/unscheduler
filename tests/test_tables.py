import tables
from pathlib import Path
import unittest, collections
from building import Building, Story
from land import Lot, Subplot

class TestTable(unittest.TestCase):

    def test_latex_factory(self):
        template = '{}'
        matrix = ['a b'.split(), '1 2'.split()]
        headers = True
        title = False
        t = tables.LaTexFactory.get_latex(matrix, template, headers=headers, title=title)
        self.assertTrue(type(t) == str)

    def test_story_table_factory(self):
        s1 = Story(0, 'terreo', 1, 1)
        t = tables.StoryTableFactory.get_latex(s1)
        self.assertTrue(type(t) == str)

    def test_building_table_factory(self):
        s1 = Story(0, 'terreo', 1)
        s2 = Story(1, 'sup', 1, 1)
        building = Building('test', [s1, s2])
        t = tables.BuildingTableFactory.get_latex(building)
        self.assertTrue(type(t) == str)
        
    def test_suplot_areas_factory(self):
        s1 = Story(0, 'terreo', 1)
        s2 = Story(1, 'sup', 1, 1)
        building = Building('test', [s1, s2])
        subplots = [Subplot(i, str(i), 5, [building]) for i in range(2)]
        lot_info = collections.defaultdict(lambda : int(10))
        lot = Lot(subplots, lot_info)
        t = tables.SubplotAreasFactory.get_latex(lot)

    def test_subplot_stats_factory(self):
        s1 = Story(0, 'terreo', 1)
        s2 = Story(1, 'sup', 1, 1)
        building = Building('test', [s1, s2])
        subplots = [Subplot(i, str(i), 5, [building]) for i in range(2)]
        lot_info = collections.defaultdict(lambda : int(10))
        lot = Lot(subplots, lot_info)
        t = tables.SubplotStatsFactory.get_latex(lot)

    def test_lot_stats_factory(self):
        s1 = Story(0, 'terreo', 1)
        s2 = Story(1, 'sup', 1, 1)
        building = Building('test', [s1, s2])
        subplots = [Subplot(i, str(i), 5, [building]) for i in range(2)]
        lot_info = collections.defaultdict(lambda : int(10))
        lot = Lot(subplots, lot_info)
        t = tables.LotStatsFactory.get_latex(lot)

  
if __name__ == '__main__':
    unittest.main()
