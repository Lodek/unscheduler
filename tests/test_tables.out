......
----------------------------------------------------------------------
Ran 6 tests in 0.007s

OK
